hushline/model.py

```diff
@@ -39,72 +75,129 @@ class User(Model):
     )
 
     @property
-    def password_hash(self) -> str:
+    def _aad_secret(self) -> bytearray:
+        domain = b"user_aad_secret"
+        aad = deque([self.__tablename__.encode()])
+        vault = current_app.config["VAULT"]
+        return bytearray(vault.decrypt(self._hidden_aad_secret, domain=domain, aad=aad))
+
+    @_aad_secret.setter
+    def _aad_secret(self, value: bytes) -> None:
+        domain = b"user_aad_secret"
+        aad = deque([self.__tablename__.encode()])
+        vault = current_app.config["VAULT"]
+        self._hidden_aad_secret = vault.encrypt(value, domain=domain, aad=aad)
+
+    @property
+    def password_hash(self) -> str | None:
         """Return the hashed password."""
         return self._password_hash
 
     @password_hash.setter
     def password_hash(self, plaintext_password: str) -> None:
         """Hash plaintext password using scrypt and store it."""
-        self._password_hash = scrypt.hash(plaintext_password)
+        domain = b"user_password_hash"
+        vault = current_app.config["VAULT"]
+        self._password_revision_number += 1
+        with temp_user_aad(self) as aad:
+            aad.append(self._password_revision_number.to_bytes(8, "big"))
+            aad.append(bytearray(plaintext_password, encoding="utf-8"))
+            # General Warnings & Guidelines:
+            # https://cheatsheetseries.owasp.org/cheatsheets/Password_Storage_Cheat_Sheet.html
+            #
+            # This pre-hash isn't vulnerable to password shucking due to the high-entropy and/or
+            # user-specific `aad` components, the admin secret, & the other domain separators.
+            # Reference:
+            # https://www.youtube.com/watch?v=OQD3qDYMyYQ
+            #
+            # This 32-byte pre-hash is also hex-normalized, which avoids implementation-specific
+            # hazards in some password hashing algorithms over raw bytes, & too long passwords.
+            # Reference:
+            # https://blog.ircmaxell.com/2015/03/security-issue-combining-bcrypt-with.html
+            pre_hashed_password = vault._derive_key(domain=domain, aad=aad)
+            self._password_hash = scrypt.hash(pre_hashed_password.hex())
 
     def check_password(self, plaintext_password: str) -> bool:
         """Check the plaintext password against the stored hash."""
-        return scrypt.verify(plaintext_password, self._password_hash)
+        domain = b"user_password_hash"
+        vault = current_app.config["VAULT"]
+        with temp_user_aad(self) as aad:
+            aad.append(self._password_revision_number.to_bytes(8, "big"))
+            aad.append(bytearray(plaintext_password, encoding="utf-8"))
+            pre_hashed_password = vault._derive_key(domain=domain, aad=aad)
+        return scrypt.verify(pre_hashed_password.hex(), self.password_hash)
```

Comment:
Switch from scrypt to the preferred / recommended argon2.

Reply:
Addressed by (https://github.com/scidsg/hushline/pull/411/commits/6f53d35d4a1c4d29292f3bb7c093187192a02ea9)